### ISSUE TITLE
Remove large objects from repository

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#2890]
Continuing a discussion began in #2882...

It was noted that the size of this repo is nearing the soft upper bound enforced by GitHub (1 GiB).  It was suggested that we rewrite history to remove large binary files from this repo (either permanently or move them to LFS).

Here's a list of the top 20 objects in the repo:

Size (KiB) | SHA | Location
--: | :-- | :--
42951 | e7f384b9de710921dacc5d737fdcbb7df3817323 | installer/jre7_21_x86.zip
42863 | 38cdc6722bcd37e3e14358d3f2ce50f5d51e53a5 | installer/jre7_u72_x86.zip
42612 | 7a307505a4ce48f47bc132bf676f527e700b707d | installer/jre7_u51_x86.zip
31831 | 05a8dfb4abdfd69c818df426e90d059a5668a2f2 | installer/jre1.6.0_25.zip
30286 | 6cc300e3b0aaf746c89b610695418262825d4935 | installer/jre1.6.0_33_x86.zip
23808 | 310e1c9dad41a063ecdc59d3e7df2e8ba0f4b05e | installer/jre1.6.0_11.zip
22607 | 38b250ebe2c9a974846ba5765c8b9a37b335e9ee | installer/jre1.6.0_03.zip
20190 | 6894f10ed4e2fe93b9b1b2b10f9b2b7cba37a186 | TripleA_RuleBook.pdf
19699 | 64312b511a2a0a42a0a56315274a7e78465f84c5 | installer/jre1.5.0_06.zip
19107 | 29fdda309360c8ed8239ba882014d87dacfa1ca7 | installer/jre1.5.0_09.zip
14885 | e72317d20ca23669ec88a99038fdb9a08c5e7a70 | assets/sounds/preindustrial/game_won/game_won_01_end_of_1812_overture.wav
13113 | 419885bfcbb49ac275491653152d9506fc532ee6 | assets/sounds/misc/game_start/game_start_02_serious_epic_theme.wav
11449 | b475dac0e8cd15f3dbd2af363c29466f00056e58 | TripleA_RuleBook.docx
10329 | c789a6a749794327f6512a7969cddcda38f087f3 | old/triplea_1_8_0_9.jar
7515 | b770e3da68611af51e8fcd278869317d24c5d827 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
6638 | 46ec6ef2cf0a7e073923596e0d31285b47e00d02 | old/triplea_1_7_0_5.jar
6432 | 3b6ee0cc5165b121e770bd704093001ffac88dda | assets/sounds/preindustrial/game_start/game_start_01_start_of_1812_overture.wav
6299 | a5b641aecf2ed535ce8d4c323be85c511170d33b | old/triplea_1_6_1_5.jar
5883 | 59ba28278fc4c66bf9b640f9af56ab49d2d19a09 | additonalImageData/relief.png
4761 | f6b754209d1c8b36c3033955f11aa5cac1b99032 | src/main/resources/org/triplea/client/ui/javafx/images/wood.png

The JREs and rule book documents are the most obvious candidates for permanent removal, as their absence should not affect rebuilding an earlier version of TripleA.

---

#### Files to remove

* [_TripleA_RuleBook.*_](https://github.com/triplea-game/triplea/files/1701501/delete-triplea-rulebook.txt)
* [_installer/**/*_](https://github.com/triplea-game/triplea/files/1701504/delete-installer.txt)
* [_old/**/*_](https://github.com/triplea-game/triplea/files/1701507/delete-old.txt)
* [_**/*.wav_](https://github.com/triplea-game/triplea/files/1701508/delete-wav.txt)

#### Procedure

1. Backup all files targeted for removal (**destination: TBD**).
1. Close or merge all open PRs.
1. Delete or merge all feature branches.
1. Have at least three devs create mirrors of the repo.
    * `git clone --mirror git://github.com/triplea-game/triplea.git triplea-backup.git`
1. Run script to remove large files from the repo ([source](https://gist.github.com/ssoloff/10693e2d5e81c45d1d8d3c27733b0a04)).
<b>:warning: DANGER ZONE :warning:</b>
1. Disable `master` branch protection.
1. Force push changes from (5).
1. Re-enable `master` branch protection.

**Comment by RoiEXLab** [comment]
Just wanted to mention #2892 does not solve any of those issues here.

**Comment by ron-murhammer** [comment]
Yeah, probably removing the old JREs would be a good step and save like 200MB.

**Comment by DanVanAtta** [comment]
Everything except for images is pretty safely removed.

@ssoloff would you be willing to do the dark git-fu on this? AFAIK we need a history rewrite and force-push. My comp/OS is really unhappy when history replays the .gitignore file LF settings; that is what stopped me when previously looking at this. We can also even remove all *wav files, we only read mp3 now.

**Comment by RoiEXLab** [comment]
Before doing the history rewrite:
I'd appreciate if the then permanently removed files could be backed up in a zip or 7zip file in case file size is no longer an issue in the future. 
Noting that some of the JavaFX images are tracked by LFS now, but the original image is still in the git history somewhere. Git LFS has a tool for that I believe. 

[The typical "make a backup" notice (Although we all should have the current version locally)]

Also noting branch protection needs to be disabled by @ron-murhammer for this particular step I believe.

**Comment by ssoloff** [comment]
Sure, I can test it out on a clone to make sure we are aware of all potential issues.  GitHub has a pretty good [write-up](https://help.github.com/articles/removing-sensitive-data-from-a-repository/) that calls out several of those issues (e.g. open PRs or unmerged branches basically becoming useless).

**Comment by ssoloff** [comment]
I added a list to the issue description containing the suggested files to remove so far.

I used the BFG Repo-Cleaner tool on a clone of the repo to remove those files, expired everything in the reflog, and performed a GC.  The bare repo size went from 748 MiB to 297 MiB, a decrease of 60%.  Not bad.

Here's the updated list of the top 20 largest objects in the repo after removing the previously-mentioned files:

Size (KiB) | Packed Size (KiB) | Location
--: | --: | :--
12565 | 11278 | old/triplea_1_9_0_0.jar
10329 | 9308 | old/triplea_1_8_0_9.jar
7515 | 7514 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
6638 | 6049 | old/triplea_1_7_0_5.jar
6299 | 5749 | old/triplea_1_6_1_5.jar
5883 | 5878 | additonalImageData/relief.png
4761 | 4762 | src/main/resources/org/triplea/client/ui/javafx/images/wood.png
3794 | 3794 | additonalImageData/classic/relief.png
3688 | 3688 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
3228 | 2996 | ant/findbugs-1.3.9/lib/findbugs.jar
2903 | 2680 | old/triplea_1_9_0_0.jar
2764 | 2541 | lib/derby-10.10.1.1.jar
2705 | 128 | test_data/Total_World_War_Dec1941_test.xml
2561 | 2468 | old/triplea_1_5_2_2.jar
2282 | 118 | src/test/resources/Total_World_War_Dec1941.xml
2203 | 1948 | lib/guava-18.0.jar
2081 | 1916 | lib/derby_10_1_2.jar
2080 | 2005 | old/triplea_1_6_0_1.jar
2055 | 2055 | src/main/resources/org/triplea/client/ui/javafx/images/wood.png
1476 | 1310 | lib/substance-7.1.jar

---

**Do we also want to remove all old jar files (i.e. _old/\*\*/*_)?**  That would get the bare repo size down to 262 MiB.

**Comment by RoiEXLab** [comment]
@ssoloff No objections on removing all old jar files as long as a backup just like on any other "original file" is being made that will be kept around.
By "original file" I mean stuff like old releases and for example wav files that have been lossily compressed to save disk space.

**Comment by ssoloff** [comment]
BTW, where shall we back up the files we're going to delete?  On one of the Linode servers?  Or should we create a new repo for "archived" files?

**Comment by RoiEXLab** [comment]
IMO it's sufficient to store a zip file on a normal cloud that's publicly available.

**Comment by DanVanAtta** [comment]
'archived' file, zip file, i'm okay either way. I'm also okay with the files being dropped entirely. I personally would probably never look to the git history to find an old-jar file or a wav file. For the wav files we have them checked in as mp3 (good candidates for LFS maybe?). The old-jar we should have available still on the old SVN page. We maintain a list of old releases on website: http://triplea-game.org/old_downloads/, regardless of what we do here we should add those files to that list for completeness sake.

**Comment by ssoloff** [comment]
@DanVanAtta The only reason I can think of needing the deleted files is to recreate an older build.  Not sure how much priority we need to give to that "requirement."

> good candidates for LFS maybe?

I think LFS usage is on hold indefinitely due to insufficient bandwidth quota on the free plan.  See #2921.

**Comment by DanVanAtta** [comment]
> The only reason I can think of needing the deleted files is to recreate an older build. 

I put almost no weight on this. We've got github releases with the artifacts, that goes back as far as we need.

**Comment by ssoloff** [comment]
It sounds like there's agreement on removing _old/**/*_, as well, so I've added that entry to the list.

I updated the description with a basic procedure we'll need to follow to accomplish this task.  Please advise if I missed anything.  Two items are still incomplete:

* I'd appreciate concrete suggestions for (1).  I don't want to upload ~1 GiB worth of files only to be told the service I chose might not be appropriate. :smile:  I'd still prefer just to create a new repo under the triplea-game organization called "old_large_files" or something, so it can be managed by the team.
* I'm going to write a gist for the script described in (5) so everyone can review and sign-off on it before it gets run.

**Comment by ssoloff** [comment]
> The old-jar we should have available still on the old SVN page.

@DanVanAtta Would that be [here](https://sourceforge.net/projects/triplea/files/)?  If so, all files appear to have been deleted.  Is there a different location that would contain all the old jars, or will they have to be backed up during step (1) of the procedure, as well?

**Comment by ssoloff** [comment]
Just for completeness, here's the top 40 objects in the repo after removing _old/**/*_.

Does anyone see any other candidates for removal?  If we don't need to recreate old builds, the _ant/_ and _lib/_ folders can probably be removed.  Also, I'm assuming the content of the _maps/_ folder has been moved to the various map repos, and is probably no longer necessary.

Size (KiB) | Packed Size (KiB) | Location
--: | --: | :--
7515 | 7514 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
5883 | 5878 | additonalImageData/relief.png
4761 | 4762 | src/main/resources/org/triplea/client/ui/javafx/images/wood.png
3794 | 3794 | additonalImageData/classic/relief.png
3688 | 3688 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
3228 | 2996 | ant/findbugs-1.3.9/lib/findbugs.jar
2764 | 2541 | lib/derby-10.10.1.1.jar
2705 | 128 | test_data/Total_World_War_Dec1941_test.xml
2282 | 118 | src/test/resources/Total_World_War_Dec1941.xml
2203 | 1948 | lib/guava-18.0.jar
2081 | 1916 | lib/derby_10_1_2.jar
2055 | 2055 | src/main/resources/org/triplea/client/ui/javafx/images/wood.png
1700 | 1533 | lib/substance-5.3.jar
1476 | 1310 | lib/substance-7.1.jar
1452 | 1288 | lib/swingx-all-1.6.4.jar
1372 | 64 | test_data/ww2_g40_balanced.xml
1259 | 1163 | ant/findbugs-1.3.9/lib/ant.jar
1255 | 1122 | lib/substance-7.3-lite.jar
1242 | 1106 | lib/substance-7.2.1-lite.jar
1205 | 1051 | lib/mockito-all-1.10.19.jar
1019 | 280 | maps/Napoleonic_Empire-Political_Map_Skin/polygons.txt
856 | 857 | src/main/resources/org/triplea/client/ui/javafx/images/background.png
797 | 214 | maps/great_war/polygons.txt
758 | 759 | additonalImageData/classic/largeMap.gif
731 | 204 | maps/great_war/polygons.txt
698 | 191 | maps/new_world_order/polygons.txt
689 | 198 | maps/big_world/polygons.txt
613 | 578 | lib/substance_5.2.jar
605 | 174 | maps/middle_earth-map_skin2/polygons.txt
604 | 604 | doc/images/map_skin_1.png
599 | 569 | lib/substance.5.1.jar
583 | 513 | lib/flamingo-7.3.jar
580 | 510 | lib/flamingo-7.2.1.jar
578 | 163 | maps/aa50/polygons.txt
562 | 158 | data/games/strategy/triplea/ui/new_polygons.txt
550 | 550 | doc/images/map_skin_2.png
525 | 525 | doc/images/map_skin_1.png
522 | 470 | ant/findbugs-1.3.9/lib/bcel.jar
516 | 444 | maps/middle_earth/lotr_middle_earth_base_image.pdn
507 | 140 | maps/europe/polygons.txt

**Comment by ron-murhammer** [comment]
@ssoloff Can you update your initial post to list all of the files to be removed? I'm fine with removing the JREs and old TripleA JARs. Is the rulebook located somewhere else now? Are those sound files being used?

**Comment by ssoloff** [comment]
@ron-murhammer

> Can you update your initial post to list all of the files to be removed?

Sure.  I think BFG Repo-Cleaner logs the files it deletes.  Let me find them, and I'll update the description.

> Is the rulebook located somewhere else now?

It's over in the [assets repo](https://github.com/triplea-game/assets/tree/master/rulebooks).

> Are those sound files being used?

According to @DanVanAtta, _.wav_ files are no longer used; all sounds are in _.mp3_ files.  I did a quick `find . -name '*.wav'` in `master` and got zero hits.

**Comment by ssoloff** [comment]
> Can you update your initial post to list all of the files to be removed?

@ron-murhammer So the BFG logs aren't very useful.  They only provide the file SHA and name, but no path.  I suppose I could pump the SHAs back through `git rev-list` to extract the path.

Also, the logs only contain the deleted files when I ran BFG with the `--delete-files` flag.  For the cases where I used `--delete-folders`, it doesn't provide the list of files actually deleted.

I tried using `git log --diff-filter=D --summary | grep delete | grep <path>` to get a list of deleted files at the specified path that _should_ be affected by BFG.  However, that doesn't seem to provide the full picture.  When I ran it using a path of `installer/`, it left out many of the zipped JREs that were reported by other tools.

I'll need to do further research to see if I can get a complete list of files that would be deleted.

**Comment by DanVanAtta** [comment]
> @DanVanAtta Would that be here? If so, all files appear to have been deleted. Is there a different location that would contain all the old jars, or will they have to be backed up during step (1) of the procedure, as well?

They're in a slightly different location: https://sourceforge.net/p/triplea/code/HEAD/tree/trunk/triplea/old/  (along with all the original wav files as well)

**Comment by DanVanAtta** [comment]
Since the SVN migration to github, I can't think of any binary files added in since then that we would drop from history and still want to keep. AFAIK anything we would drop would still be in SVN.

**Comment by ssoloff** [comment]
@ron-murhammer Ok, I managed to generate the list of files to be deleted.  For each entry in the **Files to remove** list in the description, I added a link to a file that contains the list of files that will be deleted for the entry's pattern.

For reference, I generated these lists using a command similar to the following:

```
$ git rev-list --all --objects | cut -f 2 -d ' ' | grep -P '\.wav$' | sort | uniq
```

Note that these lists should be accurate, but they are technically an approximation since BFG Repo-Cleaner almost assuredly uses a different algorithm to do path matching.
